File: 0147-insertion-sort-list/0147-insertion-sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        prev, curr, nxt = dummy, head, None
        while curr:
            nxt = curr.next
            while prev.next and prev.next.val < curr.val:
                prev = prev.next
            curr.next = prev.next
            prev.next = curr
            prev = dummy
            curr = nxt
        return dummy.next

File: 0147-insertion-sort-list/0147-insertion-sort-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var insertionSortList = function (head) {
  const dummy = new ListNode();
  let [prev, curr, next] = [dummy, head, null];

  while (curr) {
    next = curr.next;
    while (prev.next?.val < curr.val) prev = prev.next;

    curr.next = prev.next;
    prev.next = curr;
    prev = dummy;
    curr = next;
  }

  return dummy.next;
};
